### PR TITLE
parser: make the prepared statement parameter marker support escape

### LINKED ISSUE #32246: TiDB is not allowing Prepared statement parameter as argument to ESCAPE
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t1;
create table t1 (a varchar(10), key(a));
insert into t1 values ('a'), ('a\\b');
prepare stmt1 from 'select * from t1 where a like \'a\\%\' escape ?';
set @esc='#';
execute stmt1 using @esc;  
deallocate prepare stmt1;

```

### 2. What did you expect to see? (Required)
```sql
mysql> prepare stmt1 from 'select * from t1 where a like \'a\\%\' escape ?';
Query OK, 0 rows affected (0.00 sec)
Statement prepared
```
### 3. What did you see instead (Required)
```sql
mysql> prepare stmt1 from 'select * from t1 where a like \'a\\%\' escape ?';
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 44 near "?" 
mysql> set @esc='#';
Query OK, 0 rows affected (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.5.0-alpha-182-gad9430039
Edition: Community
Git Commit Hash: ad9430039f54bb9af78d44831c176bc5eafcbba0
Git Branch: master
UTC Build Time: 2022-02-04 18:58:58
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by fanrenhoo**
Signed-off-by: fanrenhoo <fanrenhoo@sina.com>

### What problem does this PR solve?

Issue Number: close #32246

Problem Summary: the prepared statement did not support escape to have a paramenter marker ?, parser will treat it as wrong syntax as normal sql

### What is changed and how it works?
I added the paramMarker in parser.y to support "ESCAPE paramMarker", so if in the prepared statement, now we can write "escape ?", that mean the escape character can also be parameterized in prepared statement. 
Add also add a test case for this
I did not commit the parser.go, because I think it's should be auto generated from parser.y

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
I tested, now it works:

>drop table if exists t1;
>create table t1 (a varchar(10), key(a));
>prepare stmt1 from 'select * from t1 where a like \'a\\%\' escape ?';
Query OK, 0 rows affected (0.00 sec)

- [ ] No code

Side effects
No
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation
No
- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Make the prepared statement support escape character be parameterized(issue 32246). the prepared statement "escape ?", , parser will treat it as wrong syntax as normal sql, now the escape character can also be parameterized in prepared statement.
```
/cc @xiongjiwei

### CODE REVIEW

**Comment by xiongjiwei** [comment]
> it's should be auto generated from parser.y

it won't. 

btw, this pr only fixed the parser part, seems it will not work as expected. Can you test the case in the issue using this PR?

**Comment by fanrenhoo** [comment]
> > it's should be auto generated from parser.y
> 
> it won't.
> 
> btw, this pr only fixed the parser part, seems it will not work as expected. Can you test the case in the issue using this PR?

OK

**Comment by fanrenhoo** [comment]
> > > it's should be auto generated from parser.y
> > 
> > 
> > it won't.
> > btw, this pr only fixed the parser part, seems it will not work as expected. Can you test the case in the issue using this PR?
> 
> OK

I checked and tested, sure this fix is not complete or not right, cause the ParamMarkerExpr is not generated in the generated ast stmt, so ParamMarkerExtractor could not find "?" afterwards. 
Let me try again on this, what I doubted is: it seems i need to add another PatternLikeExpr in parser/ast/expressions.go? but i thought this should can be fixed just in parser/parser.y, but current PatternLikeExpr already define Escape as byte, cannot save ParamMarkerExpr as ExprNode. let me  research the try more time. or may can give a bit hint THKS

**Comment by xiongjiwei** [comment]
btw, I do not think this is an easy issue for a contributor who is new to tidb😂

**Comment by fanrenhoo** [comment]
> 

You should be right. What I am trying is like you said change the PatternLikeExpr's Escape as ExprNode, but seem Accept, Restore, Format all need change. like in Restore the escape := string(n.Escape) I need to use ExprNode to transfer. It true not easy for me _^^_, let me try again. And Thks very much!

**Comment by fanrenhoo** [comment]
> > 
> 
> You should be right. What I am trying is like you said change the PatternLikeExpr's Escape as ExprNode, but seem Accept, Restore, Format all need change. like in Restore the escape := string(n.Escape) I need to use ExprNode to transfer. It true not easy for me _^^_, let me try again. And Thks very much!

I did change the PatternLikeExpr's Escape as ExprNode, and refer LimitOption's syntax, modified some files before has the byte Escape to like "n.Escape.(ast.ValueExpr).GetString()[]". Now It seem works with Parser, got ParamExtractorMarker correct in stmt. But execute with error message "ERROR 1105 (HY000): runtime error: index out of range [0] with length 0"
I traced found in executor/prepared.go, when call NormalizeDigest() with "select * from t1 where a like 'a\' escape ?", will be Normalized as "select * from t1 where a like ? escape ?", I'm not sure if this is right with normalize(seem in comment normalize will replace sth with ?)?

**Comment by fanrenhoo** [comment]
> > > 
> > 
> > 
> > You should be right. What I am trying is like you said change the PatternLikeExpr's Escape as ExprNode, but seem Accept, Restore, Format all need change. like in Restore the escape := string(n.Escape) I need to use ExprNode to transfer. It true not easy for me _^^_, let me try again. And Thks very much!
> 
> I did change the PatternLikeExpr's Escape as ExprNode, and refer LimitOption's syntax, modified some files before has the byte Escape to like "n.Escape.(ast.ValueExpr).GetString()[]". Now It seem works with Parser, got ParamExtractorMarker correct in stmt. But execute with error message "ERROR 1105 (HY000): runtime error: index out of range [0] with length 0" I traced found in executor/prepared.go, when call NormalizeDigest() with "select * from t1 where a like 'a' escape ?", will be Normalized as "select * from t1 where a like ? escape ?", I'm not sure if this is right with normalize(seem in comment normalize will replace sth with ?)?

Sorry, I think the normalize to as "select * from t1 where a like ? escape ?" may correct, the problem should when try to build plan with plannercore with the escape paramMarker. the plan builder for this case also need to be modified?

**Comment by xiongjiwei** [comment]
you need also to commit `parser.go`, and see the CI, this pr without `parser.go` will build failed

**Comment by fanrenhoo** [comment]
> you need also to commit `parser.go`, and see the CI, this pr without `parser.go` will build failed

I committed again with parser'parser.go, before i was afraid make the repo mess cause it should be auto-generated. Please review and THSK.

**Comment by xiongjiwei** [comment]
please see the failed CI, there are some runtime error, please fix it

**Comment by fanrenhoo** [comment]
> please see the failed CI, there are some runtime error, please fix it

seems some logic need be complete. OK

**Comment by fanrenhoo** [comment]
> > please see the failed CI, there are some runtime error, please fix it
> 
> seems some logic need be complete. OK

I need to check more, it seem the "select xx like" statement also affected

**Comment by fanrenhoo** [comment]
Hi, I made the new fix and commited, tested on my machine these cases "prepare 'select xxx from xxx where x like x escape xxx'", "show tables like xxx" "select xxx like x from xxx" should be right now. What need sure is：to fix "select xxx like x from xxx"(Column field has like expr) case, in "planner/core/expression_rewriter.go" change the hardcode number(previous code is also hardcode, so seems fine) in func patternLikeToExpression for er.ctrStack[] cursor, because I found when i change PatternLikeExpr.Escape as ExprNode, there is one more expression added to er.ctrStack to number 3(before is 2). I tested previous error gone but seem some result error in CI. I'll check more to see and may some hint helpful

**Comment by fanrenhoo** [comment]
the submit has problem, let me check again, sorry

File: parser/ast/expressions.go [hunk 1]
```diff
@@ -905,7 +905,7 @@ func (n *PatternLikeExpr) Restore(ctx *format.RestoreCtx) error {
 		return errors.Annotate(err, "An error occurred while restore PatternLikeExpr.Pattern")
 	}
 
-	escape := string(n.Escape)
+	escape := string(n.Escape.(ValueExpr).GetString()[0])
```
Comment (xiongjiwei) [inline parser/ast/expressions.go:908]: `escape` is ExprNode now, we cannot assume it is a `valueExpr`. see how `LimitClause` restore.

File: parser/parser.y [hunk 2]
```diff
@@ -5412,18 +5413,11 @@ PredicateExpr:
...
-			return 1
-		} else if len(escape) == 0 {
-			escape = "\\"
-		}
 		$$ = &ast.PatternLikeExpr{
 			Expr:    $1,
 			Pattern: $3,
 			Not:     !$2.(bool),
-			Escape:  escape[0],
+			Escape:  $4.(ast.ValueExpr),
```
Comment (xiongjiwei) [inline parser/parser.y:5420]: we cannot assume $4 is a `ValueExpr`, it may also be `ParamMarkerExpr`

File: planner/core/expression_rewriter.go [hunk 3]
```diff
@@ -1632,34 +1632,34 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
...
+				function, er.err = er.constructBinaryOpFunction(er.ctxStack[l-3],
 					&expression.Constant{Value: types.NewStringDatum(string(patValue)), RetType: fieldType},
 					op)
 				isPatternExactMatch = true
 			}
 		}
 	}
 	if !isPatternExactMatch {
-		types.DefaultTypeForValue(int(v.Escape), fieldType, char, col)
+		types.DefaultTypeForValue(int(v.Escape.(ast.ValueExpr).GetString()[0]), fieldType, char, col)
```
Comment (xiongjiwei) [inline planner/core/expression_rewriter.go:1657]: same

File: planner/core/planbuilder.go [hunk 4]
```diff
@@ -2842,6 +2842,8 @@ func splitWhere(where ast.ExprNode) []ast.ExprNode {
 		}
 	case *ast.ParenthesesExpr:
 		conditions = append(conditions, splitWhere(x.Expr)...)
+	case *ast.PatternLikeExpr:
```
Comment (xiongjiwei) [inline planner/core/planbuilder.go:2845]: `splitWhere` is used for `select * from t where a = 1 and b=2`, it will split `a=1 and b=2` into a list that contains `[a=1, b=2]`, so I think no need change this

**Comment by xiongjiwei** [comment]
please refer `LimitClause`, I think I cannot give you more advice because I am not familiar with it either, but I think we can follow how the `LimitClause` is done strictly.

**Comment by fanrenhoo** [comment]
> please refer `LimitClause`, I think I cannot give you more advice because I am not familiar with it either, but I think we can follow how the `LimitClause` is done strictly.

OK. thks

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi，I committed new fixes. as suggest, to learn what LimitClause does as I could. I tested with as much cases as I can find seems work now. And PatternLikeExpr with escape can't be all same as Limit because a LikeEscapeOpt there so can't create a brand new "EscapeClause", so still some other places changes.  BTW, the ParamMarkerExpr is derived from ValueExpr, so it should can use it in parser(LimitClause seems like so also).
Please check, THKS

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi，I committed new fixes. as suggest, to learn what LimitClause does as I could. I tested with as much cases as I can find seems work now. And PatternLikeExpr with escape can't be all same as Limit because a LikeEscapeOpt there so can't create a brand new "EscapeClause", so still some other places changes. BTW, the ParamMarkerExpr is derived from ValueExpr, so it should can use it in parser(LimitClause seems like so also).
And its still to much modified for new TiDB contributor, I am not sure if some other cases will be affected.
Please check, THKS

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi，I committed new fixes. as suggest, to learn what LimitClause does as I could. I tested with as much cases as I can find seems work now. And PatternLikeExpr with escape can't be all same as Limit because a LikeEscapeOpt there so can't create a brand new "EscapeClause", so still some other places changes. BTW, the ParamMarkerExpr is derived from ValueExpr, so it should can use it in parser(LimitClause seems like so also). And the PatternLikeToExpression's hardcode "l-.." looks ugly but I just try to follow the previous logic
And its still to much modified for new TiDB contributor, I am not sure if some other cases will be affected.
Please check, THKS

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi，I committed new fixes. as suggested, to learn from what LimitClause does as I could. I tested as much cases as I can find now seems work fine now. And PatternLikeExpr with escape can't be all same as Limit because a LikeEscapeOpt there so can't create seperate "EscapeClause", so still some other places changes. BTW, the ParamMarkerExpr is derived from ValueExpr, so it should can use it in parser(LimitClause is like so also). And the PatternLikeToExpression's hardcode "l-.." looks ugly but I just try to follow the previous logic. It looks work fine now, still need to check to make sure.
Please check, and THKS for guide and help.

**Comment by xiongjiwei** [comment]
> And the PatternLikeToExpression's hardcode

it's not hardcode, there is a stack to hold the arguments about the function. For `like` function, there may be 3 args, 

```
select 'a' like 'a' escape '\\'
        ^1       ^2         ^3
```
and we push these 3 args to the argument stack, for the original code, the 3rd argument is always a byte, we can skip this step, just using the value in `ast` is fine. But after we support the parameter maker, this argument will also be pushed into the stack, so the number is changed

**Comment by xiongjiwei** [comment]
please also confirm the CI is passed. CI will run after your every commit.

File: parser/parser.y [hunk 5]
```diff
@@ -5439,11 +5441,23 @@ RegexpSym:
...
+		$$ = expr
 	}
-|	"ESCAPE" stringLit
+|	"ESCAPE" EscapeOption
 	{
-		$$ = $2
+		$$ = $2.(ast.ValueExpr)
+	}
+
+EscapeOption:
```
[the remark this answers is not in the document]
Comment (fanrenhoo) [reply parser/parser.y:5452]: THKS much! now understand what the '...Option' means, but with the 'limit', LimitOption in parer.y is still defined like this. Do I need to ignore the Limit thing, just change the EscapeOption to EscapeExpr?

[the remark this answers is not in the document]
Comment (xiongjiwei) [reply parser/parser.y:5454]: just `EscapeOption` now, the limit things can be done in next PR. one PR should only force on one thing

**Comment by xiongjiwei** [comment]
and please resolve the conversations that you have handled

File: parser/ast/expressions.go [hunk 6]
```diff
@@ -923,9 +930,13 @@ func (n *PatternLikeExpr) Format(w io.Writer) {
 		fmt.Fprint(w, " LIKE ")
 	}
 	n.Pattern.Format(w)
-	if n.Escape != '\\' {
+	escape := "?"
```
[the remark this answers is not in the document]
Comment (fanrenhoo) [reply parser/ast/expressions.go]: Hi, I made new modification as suggested, in parser.y, PatternLikeExpr "Escape: $4.(ast.ValueExpr)" still there cause seem compile error happened, and like this look also ok for later.

**Comment by fanrenhoo** [comment]
I commited with new changes as suggested.  I see CI some error report but not sure related will the change. please check, THKS

**Comment by fanrenhoo** [comment]
still some problem, I'll check more

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi, I made changes with suggest and committed, now it looks fine. what need to say is in "parser/ast/expressions.go" PatternLikeExpr Restore and Format, for sure use "Escape.Restore" "Escape.Format" is nicer, but I found parser_test and ast/format_test show errors, I guess the ValueExpr's restore/format not compatible with the "_test.go" already there, may can modify the "_test" instead, but theres is also a syntax error came with escape(_UTF8MB4). So for safe I keep it strictly same as the previous hard format, but keep the "if _, ok := n.Escape.(ParamMarkerExpr); !ok {" which is much better than before.
please review, THKS much.

**Comment by fanrenhoo** [comment]
also in parser.y LikeEscapOpt, I added ' escape '' ' case to complete logic, and EscapeOption now is EscapeExpr also.

File: parser/parser.y [hunk 7]
```diff
@@ -5412,18 +5413,19 @@ PredicateExpr:
...
+                        	yylex.AppendError(ErrWrongArguments.GenWithStackByArgs("ESCAPE"))
+                        	return 1
+			}
 		}
 		$$ = &ast.PatternLikeExpr{
 			Expr:    $1,
 			Pattern: $3,
 			Not:     !$2.(bool),
-			Escape:  escape[0],
+			Escape:  $4.(ast.ValueExpr),
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply parser/parser.y:5429]: did not changed?

File: parser/parser.y [hunk 8]
```diff
@@ -5439,13 +5442,28 @@ RegexpSym:
...
 	{
 		$$ = $2
 	}
 
+EscapeExpr:
+	stringLit
+	{
+		if len($1) == 0 {
+			$1 = "\\"
+		}
```
[the remark this answers is not in the document]
Comment (fanrenhoo) [reply parser/parser.y:5458]: yes, from the previous code, its like this logic. and the test case give the same logic if escape '' so as none

**Comment by fanrenhoo** [comment]
> > 
> 
> you are right! I did think this before, but because the test case with "set @e = '#'“ get the right result, so I thought it should be right ? there will substitute by @e when exec stmt. Now it seems I substitute all @e with "'?'". THKS much I will check more.

Hi, I made the fix for this, I tested now it seem OK.  I guess maybe there is nicer way like in "function = er.notToExpression(...ctxStack[l-3], ctxStack[l-2], ctxStack[l-1]...", but seems not work or more to do, so I did like commited. Please check THKS much.

**Comment by xiongjiwei** [comment]
I learned how tidb handle the parameter maker, at prepare step, it will be rewritten to a function `getUserVar`, and at execute step, the variable `@xx` will be filled into the `getUserVar`, 

the `getUserVar` is already done, so I think you should only consider not `prepare stmt` scenario.

the knowledge about the expression framework in tidb is required for this issue, please let me know if you need more help

File: planner/core/expression_rewriter.go [hunk 9]
```diff
@@ -1631,35 +1631,55 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
...
+			param, err := expression.ParamMarkerExpression(er.sctx, n, false)
+			if err != nil {
+				return
+			}
+			str, isNull, err := expression.GetStringFromConstant(er.sctx, param)
+			if err != nil {
+				return
+			}
+			if isNull {
+				return
```
Comment (xiongjiwei) [inline planner/core/expression_rewriter.go:1647]: I do not think this is a good way to fix the issue

**Comment by fanrenhoo** [comment]
> know if yo

THKs very much!!! Let me deep more to find a nicer way, at least from this issue i pushed learn more knowledge about TIDB. Thks again!

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi, I found a new way to fix, mainly in planner/core/expression_rewriter.go, this should be nicer than before and treat prepare/execute in unified way, and comply with the rest code in same function. I did research and try GetUserVar also, seem this way works.  please check THKS.

File: planner/core/expression_rewriter.go [hunk 10]
```diff
@@ -1631,35 +1631,46 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
...
+			er.err = err
+			return
+		}
+		if !isNull {
+			escape = str
+		}
+	}
 	// Treat predicate 'like' the same way as predicate '=' when it is an exact match and new collation is not enabled.
-	if patExpression, ok := er.ctxStack[l-1].(*expression.Constant); ok && !collate.NewCollationEnabled() {
+	if patExpression, ok := er.ctxStack[l-2].(*expression.Constant); ok && !collate.NewCollationEnabled() {
```
Comment (xiongjiwei) [inline planner/core/expression_rewriter.go:1646]: if escape is parameter maker, we can not rewrite it to `=` either.

File: planner/core/expression_rewriter.go [hunk 11]
```diff
@@ -1631,35 +1631,46 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
...
 				types.DefaultTypeForValue(string(patValue), fieldType, char, col)
-				function, er.err = er.constructBinaryOpFunction(er.ctxStack[l-2],
+				function, er.err = er.constructBinaryOpFunction(er.ctxStack[l-3],
 					&expression.Constant{Value: types.NewStringDatum(string(patValue)), RetType: fieldType},
 					op)
 				isPatternExactMatch = true
 			}
 		}
 	}
 	if !isPatternExactMatch {
```
Comment (xiongjiwei) [inline planner/core/expression_rewriter.go:1667]: we need also to consider the parameter maker.

File: planner/core/show_predicate_extractor.go [hunk 12]
```diff
@@ -58,8 +58,12 @@ func (e *ShowColumnsTableExtractor) Extract(show *ast.ShowStmt) bool {
 		switch pattern.Pattern.(type) {
 		case *driver.ValueExpr:
 			// It is used in `SHOW COLUMNS FROM t LIKE `abc``.
+			escape := "?"
+			if _, ok := pattern.Escape.(*driver.ParamMarkerExpr); !ok {
+				escape = pattern.Escape.(*driver.ValueExpr).GetString()
+			}
```
Comment (xiongjiwei) [inline planner/core/show_predicate_extractor.go:64]: `SHOW COLUMNS FROM t LIKE abc` cannot use escape, see the parser. I think it will report a syntax error

File: executor/prepared_test.go [hunk 13]
```diff
@@ -70,6 +70,19 @@ func TestUnsupportedStmtForPrepare(t *testing.T) {
...
+	tk.MustExec("prepare stmt1 from 'select * from t1 where a like \\'a\\\\%\\' escape ?';")
+	tk.MustExec("set @e='#';")
+	result := tk.MustQuery("execute stmt1 using @e;")
+	result.Check(testkit.Rows("a\\b"))
+	tk.MustExec("deallocate prepare stmt1;")
+	tk.MustExec("prepare stmt1 from 'select * from t1 where a like \\'a\\\\%\\' escape ?';")
+	tk.MustExec("set @e='\\\\';")
+	result = tk.MustQuery("execute stmt1 using @e;")
+	result.Check(testkit.Rows())
+	tk.MustExec("deallocate prepare stmt1;")
```
Comment (xiongjiwei) [inline executor/prepared_test.go:85]: please add more tests according to my comment, there are many negative examples during this PR, please add them all.

**Comment by xiongjiwei** [comment]
please also resolve the conflict

File: planner/core/show_predicate_extractor.go (hunk 12 above)
Comment (fanrenhoo) [reply planner/core/show_predicate_extractor.go:64]: Hi, this is right from my test result, "show columns from t like abc" can run correctly, and also "show tables like ...” also run correctly

File: executor/prepared_test.go (hunk 13 above)
Comment (fanrenhoo) [reply executor/prepared_test.go:85]: OK, I will check and try to add complete test cases

File: planner/core/show_predicate_extractor.go (hunk 12 above)
Comment (fanrenhoo) [reply planner/core/show_predicate_extractor.go:64]: You are right, the prepared case for "show columns ...." syntax error. let me check more, THKS much

Comment (fanrenhoo) [reply planner/core/show_predicate_extractor.go:64]: Yes, you are right! "prepared stmt1 from 'show columns from t like abc escape ?'" and single "show columns from t like abc escape '#'" both syntax error. What I'm doubt is: this issue is about prepared stmt not compatible with regular SQL syntax, that is: select * from t1 where a like '...' escape '.' works, but prepared not works. Now the "show columns.." and "show tables..." both not works. so do I need to fix these in this issue, or should in another issue?

Comment (fanrenhoo) [reply planner/core/show_predicate_extractor.go:63]: I added logic in parser.y to make "show columns" "show tables" syntax work for escape and prepared cases. please check

File: planner/core/expression_rewriter.go (hunk 11 above)
Comment (fanrenhoo) [reply planner/core/expression_rewriter.go:1674]: I made fixed to avoid parameter marker in rewrite to "=“ logic, please check to see if its right, THKS

File: planner/core/show_predicate_extractor.go (hunk 12 above)
Comment (xiongjiwei) [reply planner/core/show_predicate_extractor.go:63]: no, keep the original logic for `show tables` and `show columns`, please follow the MySQL syntax here.

File: planner/core/expression_rewriter.go [hunk 14]
```diff
@@ -1631,35 +1631,53 @@ func (er *expressionRewriter) patternLikeToExpression(v *ast.PatternLikeExpr) {
...
+		if !isNull {
+			escape = str
+		}
+	}
 	// Treat predicate 'like' the same way as predicate '=' when it is an exact match and new collation is not enabled.
-	if patExpression, ok := er.ctxStack[l-1].(*expression.Constant); ok && !collate.NewCollationEnabled() {
+	isPreparedParamMarker := false
+	switch n := v.Escape.(type) {
+	case *driver.ParamMarkerExpr:
+		if !n.InExecute {
```
Comment (xiongjiwei) [inline planner/core/expression_rewriter.go:1649]: this file is all about transforming `AST` to `expression`, which will only happen in the build plan step, and `InExecute` is used in runtime.

Here, we should only consider the `non-paramaker` scenario, so we will try to rewrite `like` to `eq` if the escape type is `expression.constant`.

Comment (fanrenhoo) [reply planner/core/expression_rewriter.go]: Understood! I changed it as should already

**Comment by fanrenhoo** [comment]
@xiongjiwei Hi, I made new changes as suggested, I think this time it should be ok after so many times and truely learned more: change back the orginal logic of "show table/column like" as mysql, made  expression_rewriter to avoid paramMarker case only, show_predicate_extractor.go as dont need condier paramMarker at all, and rebase to solve conflict, and try add more test cases(most previous PR's negative cases actually were check out by already existed test cases, so It should enough).  Please check THKS.

**Comment by xiongjiwei** [comment]
I will review in these two days

**Comment by fanrenhoo** [comment]
> I will review in these two days

THKS

**Comment by fanrenhoo** [comment]
Hi, how about this one, THKS.

**Comment by xiongjiwei** [comment]
> Hi, how about this one, THKS.

sorry for the late reply, I will review it today or tomorrow. 😰

**Comment by xiongjiwei** [comment]
hi @fanrenhoo, I reviewed this PR today and found I may misleading you at some points, I will read the related code and give some feedback next week😰

**Comment by fanrenhoo** [comment]
> hi @fanrenhoo, I reviewed this PR today and found I may misleading you at some points, I will read the related code and give some feedback next week😰

OK. THKS

**Comment by fanrenhoo** [comment]
> hi @fanrenhoo, I reviewed this PR today and found I may misleading you at some points, I will read the related code and give some feedback next week😰

I checked again, may your previous suggestion is good and right I think._^^_

File: executor/prepared_test.go [hunk 15]
```diff
@@ -70,6 +70,29 @@ func TestUnsupportedStmtForPrepare(t *testing.T) {
...
+	tk.MustExec("prepare stmt1 from 'select * from t1 where a not like \\'a\\\\%\\' escape ?';")
+	tk.MustExec("set @e='#';")
+	result = tk.MustQuery("execute stmt1 using @e;")
+	result.Check(testkit.Rows("a"))
+	tk.MustExec("deallocate prepare stmt1;")
+	tk.MustExec("prepare stmt1 from 'select * from t1 where a not like \\'a\\\\%\\' escape ?';")
+	tk.MustExec("set @e='\\\\';")
+	result = tk.MustQuery("execute stmt1 using @e;")
+	result.Check(testkit.Rows("a", "a\\b"))
+	tk.MustExec("deallocate prepare stmt1;")
```
Comment (xuyifangreeneyes) [inline executor/prepared_test.go:95]: Can we move the test cases to a new unit function? It doesn't belong to `UnsupportedStmtForPrepare`.

File: parser/ast/expressions.go [hunk 16]
```diff
@@ -905,12 +905,16 @@ func (n *PatternLikeExpr) Restore(ctx *format.RestoreCtx) error {
...
-		ctx.WriteKeyWord(" ESCAPE ")
-		ctx.WriteString(escape)
-
+	escape := "?"
+	if _, ok := n.Escape.(ParamMarkerExpr); !ok {
+		// it can not convert to `ParamMarkerExpr`, so it must be `ValueExpr`
+		if n.Escape.(ValueExpr).GetString() == "\\" {
+			return nil
+		}
+		escape = n.Escape.(ValueExpr).GetString()
```
Comment (xuyifangreeneyes) [inline parser/ast/expressions.go:914]: Add the test case for testing `Restore`.

Comment (fanrenhoo) [reply parser/ast/expressions.go:914]: Please check, THKS.

**Comment by fanrenhoo** [comment]
@xiongjiwei @xuyifangreeneyes Hi, this PR fall back master too long time, so I create a new PR as preparedstmtescapesupport2, merged the fix to based on newest master. so please check with new PR #34039

**Comment by fanrenhoo** [comment]
Hi, how about this one, THKS

**Comment by fanrenhoo** [comment]
Hi, how about this one? its been a long time I checked again may its right I guess.

**Comment by fanrenhoo** [comment]
Hi, I tred to rebase, manunally follow steps, but many cases happened, now its seems mess. Can give me some instruction, or can I move to a brand new pull request for this issue?

**Comment by xiongjiwei** [comment]
well, the conflict file is `parser.go` it is generated by `make parser`. just re-make is fine

**Comment by fanrenhoo** [comment]
@xuyifangreeneyes hi, how about this one after follow your suggest? THKS

**Comment by fanrenhoo** [comment]
Hi, how about this one? it seem take a very long time. THKS

**Comment by fanrenhoo** [comment]
Hi, How about this one?
